Adds ability to screen Outputs by time step

jessecarterMOOSE opened the pull request:
Simple modification to check that the time step is also within a specified range, in addition to the time. If anybody has a better name for the parameter(s), I'm all ears.

Also adds a very simple test based on the existing test for output start/end time.

closes #10645

jessecarterMOOSE committed:
Adds ability to screen Outputs by time step, closes #10645

moosebuild commented:
Job [Precheck](https://www.moosebuild.org/job/151010/) on 4ce3662 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.org/docs/PRs/10646/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.org/docs/PRs/10646/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 921a67b20a7a4fd59ba20039a52240d614bc8fac`

moosebuild commented:
Job [Documentation](https://www.moosebuild.org/job/151032/) on 8df439a wanted to post the following:

View the site [here](http://mooseframework.org/docs/PRs/10646/site)

This comment will be updated on new commits.

dschwen commented on framework/src/outputs/Output.C:
```diff
@@ -114,6 +116,10 @@ Output::Output(const InputParameters & parameters)
                                            : -std::numeric_limits<Real>::max()),
     _end_time(isParamValid("end_time") ? getParam<Real>("end_time")
                                        : std::numeric_limits<Real>::max()),
+    _start_step(isParamValid("start_step") ? getParam<int>("start_step")
```
I suggest using default values instead

jessecarterMOOSE replied:
Good point. Does it always start at 0? If so, why is it not an unsigned int?

permcody replied:
See my comment above. Right now, step number is currently signed, but I don't think it needs to be. Let's leave it as signed for now and we'll roll through here later and change everything at once.

jessecarterMOOSE commented:
Looks like I made the output window cover the mid-point of the test which cases it to fail on recover. Now I see that the other tests had recover=false in the test spec. I'll add that to the new test.

permcody commented on framework/include/outputs/Output.h:
```diff
@@ -215,6 +215,12 @@ class Output : public MooseObject,
   Real _end_time;
 
+  /// Start outputting at this time step
+  int _start_step;
```
@bwspenc, @YaqiWang, This isn't @jessecarterMOOSE's issue, but do we need to support having a negative time_step number? I'm looking through MOOSE and I can't see any place where we set it negative but it's currently declared as an int. In fact, we have a few places where we have to cast it to unsigned to compare to all of the unsigned values that we use everywhere else in MOOSE. It would be nice to know if we really need this and start working on getting it out of here if we don't. 

@jessecarterMOOSE - No action on your part.

YaqiWang replied:
I do not need for now. Is this there because someone is thinking to march in time backward? We never did adjoint calculations in time, where possibly it is clearer to set dt to negative numbers. We possibly can still just increment time step from 0.

permcody replied:
Good to know, I'll open a ticket to change step over to unsigned.

permcody commented on framework/src/outputs/Output.C:
```diff
@@ -114,6 +116,10 @@ Output::Output(const InputParameters & parameters)
     _end_time(isParamValid("end_time") ? getParam<Real>("end_time")
                                        : std::numeric_limits<Real>::max()),
+    _start_step(isParamValid("start_step") ? getParam<int>("start_step")
+                                           : -std::numeric_limits<int>::max()),
```
This code can now be updated `std::numeric_limit<int>::lowest`

jessecarterMOOSE replied:
OK, I'll change this and the one above it.

permcody commented on test/tests/outputs/iterative/tests:
```diff
@@ -85,4 +85,15 @@
+    input = 'iterative_csv_check_step.i'
+    csvdiff = 'iterative_csv_check_step_out.csv'
+
+    # This test relies on number of residuals being the same, so run it serially
```
I really don't like this regression test. Number of residuals is one of the worst quantities to test for because it can change so easily (i.e. parallel, threaded, even PETSc updates). I'd suggest an even more simple test, like "simple_transient_diffusion" where we won't have to carry these restrictions.

jessecarterMOOSE replied:
I too thought that was strange, but I was just copying what had already been done. I would suggest changing all the tests here that use residual counts, but that might be better for a different PR.

jessecarterMOOSE committed:
use std::numeric_limits::lowest() in Output.C

jessecarterMOOSE commented:
@permcody I switched the test over to a simple transient test and added a couple more. This should be ready for another review. Since there was a commit in there with a test that didn't make the cut, I went ahead and fixed up the commit history.

jessecarterMOOSE commented:
Arg, I realize now that I only needed one input file and could have handled all (3) test cases with cli args.

bwspenc commented:
@permcody Regarding the negative time step, we routinely use negative times, but I don't think we ever use a negative time step.
